File: funciones.ts

type Operation = (number: number) => number;

const zero = (operation?: Operation) => !operation ? 0 : operation(0);
const one = (operation?: Operation) => !operation ? 1 : operation(1);
const two = (operation?: Operation) => !operation ? 2 : operation(2);
const three = (operation?: Operation) => !operation ? 3 : operation(3);
const four = (operation?: Operation) => !operation ? 4 : operation(4);
const five = (operation?: Operation) => !operation ? 5 : operation(5);
const six = (operation?: Operation) => !operation ? 6 : operation(6);
const seven = (operation?: Operation) => !operation ? 7 : operation(7);
const eight = (operation?: Operation) => !operation ? 8 : operation(8);
const nine = (operation?: Operation) => !operation ? 9 : operation(9);

const plus = (number: number) => (otherNumber: number) => otherNumber + number;
const minus = (number: number) => (otherNumber: number) => otherNumber - number;
const times = (number: number) => (otherNumber: number) => otherNumber * number;
const dividedBy = (number: number) => (otherNumber: number) => otherNumber / number;

console.log(one(plus(one()))); // must return 2
console.log(one(minus(zero(plus(six()))))); // must return -5

File: funciones.py
def zero(operation = None):
    return operation(0) if operation else 0

def one(operation = None):
    return operation(1) if operation else 1

def two(operation = None):
    return operation(2) if operation else 2

def three(operation = None):
    return operation(3) if operation else 3

def four(operation = None):
    return operation(4) if operation else 4

def five(operation = None):
    return operation(5) if operation else 5

def six(operation = None):
    return operation(6) if operation else 6

def seven(operation = None):
    return operation(7) if operation else 7

def eight(operation = None):
    return operation(8) if operation else 8

def nine(operation = None):
    return operation(9) if operation else 9

def plus(number):
    return lambda other_number: other_number + number
def minus(number):
   return lambda other_number: other_number - number
def times(number):
    return lambda other_number: other_number * number
def divided_by(number):
    return lambda other_number: other_number // number

print(four(times(five()))) # imprime 20
print(one(plus(eight()))) # imprime 9
print(seven(minus(three()))) # imprime 4
print(nine(divided_by(three()))) # imprime 3
